packages/cactus-api-client/src/main/typescript/socketio-api-client.ts

```diff
@@ -0,0 +1,345 @@
+import { Observable, ReplaySubject } from "rxjs";
+import { finalize } from "rxjs/operators";
+
+// TODO - Remove any from here
+export function verifyValidatorJwt(
+  keyPath: string,
+  targetData: string,
+): Promise<any> {
+  return new Promise((resolve, reject) => {
+    const publicKey = readFileSync(resolvePath(__dirname, keyPath));
+
+    const option: VerifyOptions = {
+      algorithms: ["ES256"],
+    };
+
+    verify(targetData, publicKey, option, function (err: any, decoded: any) {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(decoded);
+      }
+    });
+  });
+}
+
+export type SocketIOApiClientOptions = {
+  readonly validatorID: string;
+  readonly validatorURL: string;
+  readonly validatorKeyPath: string;
+  readonly logLevel?: LogLevelDesc;
+  readonly maxCounterRequestID?: number;
+  readonly syncFunctionTimeoutMillisecond?: number;
+  readonly socketOptions?: Partial<ManagerOptions & SocketOptions>;
+};
+
+export class SocketLedgerEvent {
+  id = "";
+  verifierId = "";
+  data: Record<string, unknown> | null = null;
+}
+
+// TODO - To entire class - analyze and handle broken connection / socket disconnected scenarios
+export class SocketIOApiClient implements ISocketApiClient<SocketLedgerEvent> {
+  private readonly log: Logger;
+  private readonly socket: Socket;
+  // TODO - Why replay only last one? Maybe make it configurable?
+  private monitorSubject: ReplaySubject<SocketLedgerEvent> | undefined;
+
+  readonly className: string;
+  counterReqID = 1;
+  checkValidator: (
+    key: string,
+    data: string,
+  ) => Promise<any> = verifyValidatorJwt;
+
+  constructor(public readonly options: SocketIOApiClientOptions) {
+    this.className = this.constructor.name;
+
+    Checks.nonBlankString(
+      options.validatorID,
+      `${this.className}::constructor() validatorID`,
+    );
+    Checks.nonBlankString(
+      options.validatorURL,
+      `${this.className}::constructor() validatorURL`,
+    );
+    Checks.nonBlankString(
+      // TODO - checks path exists?
+      options.validatorKeyPath,
+      `${this.className}::constructor() validatorKeyPath`,
+    );
+
+    const level = this.options.logLevel || "INFO";
+    const label = this.className;
+    this.log = LoggerProvider.getOrCreate({ level, label });
+
+    this.log.info(
+      `Created ApiClient for Validator ID: ${options.validatorID}, URL ${options.validatorURL}, KeyPath ${options.validatorKeyPath}`,
+    );
+    this.log.debug("socketOptions:", options.socketOptions);
+
+    this.socket = io(options.validatorURL, options.socketOptions);
+  }
+
+  public sendAsyncRequest(
+    contract: Record<string, unknown>,
+    method: Record<string, unknown>,
+    args: any,
+  ): void {
+    try {
+      const requestData = {
+        contract: contract,
+        method: method,
+        args: args,
+      };
+
+      this.log.debug("sendAsyncRequest() Request:", requestData);
+      this.socket.emit("request2", requestData);
```

Comment:
@outSH How do different types of requests get disambiguted here? Is this assumption baked in that the entire system only supports one type of request for a single use-case pertaining to a single endpoint/backend logic?

Reply:
@petermetz No typecheck here, it's assumed that the caller knows what is in the response. This is a lift-and-shift implementation from socketio-server, can be improved in the future (along with fixing other loose-typing artifacts like any/object)

Reply:
@outSH Do you mean that you'd open a brand new connection for each request/response exchange? What I mean is how would the same connection get shared between different request/response cycles?